### PR TITLE
Origin/simple calc

### PR DESCRIPTION
**Pull request by atldaveo**
Added multi-operation functionality (add/subtract/multiply/divide) using CLI. 

Summary of changes:
-CLI using flags to specify operation (-a/s/m/d) followed by two integers separated by a space
-Additional flags (requiring no arguments) to view help (-h) information and version (-v) information

###

-CLI user input flags coded via a series of switch-case statements
-Additional helper file / header added in order to support "help" and "version" displays

### CODE REVIEW

File: src/helpers.c [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#include <stdio.h>
+#include <unistd.h>
+#include "debug.h"
+
+void print_help()
+{
+    printf("Usage: simple-calc [OPTIONS]\n");
```
Comment (reecetyl) [inline src/helpers.c:7]: This is fine for now but the simplecalc spec will want the input in the form operand1 operator operand2 (like ./simplecalc 2 + 2)

File: include/math.h [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#ifndef MATH_H
+#define MATH_H
+
+int add(int x, int y);
```
Comment (reecetyl) [inline include/math.h:4]: Add doxygen style comments (there is a vscode extension to help) 

```suggestion

/**
 * @brief Adds two integers.
 *
 * This function takes two integer arguments and returns their sum.
 *
 * @param x The first integer to add.
 * @param y The second integer to add.
 * @return The sum of the two integers.
 */
int add(int x, int y);
```

Comment (reecetyl) [reply include/math.h:4]: This will also auto populate your tooltips in vscode if enabled when you hover over a call of a commented function

File: src/helpers.c (hunk 1 above)
Comment (reecetyl) [reply src/helpers.c:7]: Small nitpick but you can pass in argv[0] to this in case your executable name changes rather than hardcoding it in

File: src/main.c [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+#include <unistd.h>
+#include "math.h"
+#include "debug.h"
+#include "helpers.h"
+
+
+int main(int argc, char *argv[]) {
+    int opt, num1, num2 = 0;            // Only works for integers for now
+    char operation = 0;             // Which operator we are going to use
+    int has_numbers = 0;            // Status: Does *argv have numbers?
```
Comment (reecetyl) [inline src/main.c:12]: Since this is a true/false value a Boolean is probably better than an int

File: src/main.c [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+    // Parse command-line options
+    while ((opt = getopt(argc, argv, "a:s:m:d:hv")) != -1) // CMD Line options
```
Comment (reecetyl) [inline src/main.c:15]: Barr C will have constant before the variable in a comparison

File: src/main.c [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    int has_numbers = 0;            // Status: Does *argv have numbers?
+
+    // Parse command-line options
+    while ((opt = getopt(argc, argv, "a:s:m:d:hv")) != -1) // CMD Line options 
+    {
+        switch (opt) 
+        {
+            case 'a': // Add
+                operation = 'a';
+                num1 = atof(optarg);
```
Comment (reecetyl) [inline src/main.c:21]: 1. You can assume these will be integer not float values

2. Use strtol rather than atoi

File: src/main.c [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+                return 0;
+
+            default: // Invalid option
+                print_help();
+                return 1;
+        }
+    }
+
+    // Perform the selected operation
+    if (has_numbers)
```
Comment (reecetyl) [inline src/main.c:62]: Do you need this check if you are already returning early from help and print version?

File: src/main.c [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+    // Perform the selected operation
+    if (has_numbers) 
+    {
+        int result;
+        switch (operation) 
+        {
+            case 'a':
+                result = add(num1, num2);
+                printf("%d + %d = %d\n", num1, num2, result);
```
Comment (reecetyl) [inline src/main.c:69]: You can probably save the operator and reduce the number of printf statements (have one at the very end)

File: src/math.c [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    return x * y;
+}
+
+int divide(int x, int y)
+{
+    return x / y;
+}
+
+/*
+int add(int32_t x, int32_t y, int32_t *res) {
```
Comment (reecetyl) [inline src/math.c:26]: Next step is implement this style

File: test/include/debug.h [hunk 9]
```diff
@@ -0,0 +1,14 @@
+#ifndef DEBUG_H
```
Comment (reecetyl) [inline test/include/debug.h:1]: Seems like some duplicate files here

File: test/include/math.h [hunk 10]
```diff
@@ -0,0 +1,7 @@
+#ifndef MATH_H
```
Comment (reecetyl) [inline test/include/math.h:1]: Duplicate

File: test/src/main.c [hunk 11]
```diff
@@ -0,0 +1,13 @@
+#include <stdio.h>
```
Comment (reecetyl) [inline test/src/main.c:1]: Duplicate

File: test/src/math.c [hunk 12]
```diff
@@ -0,0 +1,8 @@
+#include "math.h"
+#include "debug.h"
```
Comment (reecetyl) [inline test/src/math.c:2]: Duplicate

File: test/test_math.c [hunk 13]
```diff
@@ -0,0 +1,56 @@
+#include <CUnit/CUnit.h>
```
Comment (reecetyl) [inline test/test_math.c:1]: Two ways you could extend this 

1. Separate file per operator
2. Separate suite per operator, and separate files for other functionality (maybe validation of user input is the only other function to test in this project)

**Review by reecetyl — CHANGES_REQUESTED** [review]

File: test/src/math.c (hunk 12 above)
Comment (atldaveo) [reply test/src/math.c:2]: Is it not standard practice to include the .c file's own header?

Comment (atldaveo) [reply test/src/math.c:2]: nvm, just saw that the test/ directory created a bunch of duplicates while I was raging at github for why it wouldn't push my previous commits

File: src/arithmetic.c [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+ * @version 2.1
+ * @date 2024-11-25
+ * 
+ */
+
+#include "arithmetic.h"
+#include "debug.h"
+
+/**
+ * @brief Addition function.
```
Comment (reecetyl) [inline src/arithmetic.c:14]: Doxygen comments usually go in header (.h) files unless they are static functions in which case they would be in the .c files. But this looks good!

File: src/arithmetic.c [hunk 15]
```diff
@@ -0,0 +1,83 @@
+/**
+ * @file arithmetic.c
```
Comment (reecetyl) [inline src/arithmetic.c:2]: Good, now put at top of every file (should be an automatic pre fill with a vscode extension)

File: src/arithmetic.c [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+
+/**
+ * @brief Division function.
+ * 
+ * @param x First number in division sequence.
+ * @param y Second number in division sequence.
+ * @return int Quotient of the argument values.
+ */
+
+int divide(int x, int y)
```
Comment (reecetyl) [inline src/arithmetic.c:61]: What if y = 0?

File: src/main.c [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+ * @param argc Number of CMD Line arguments
+ * @param argv Array of CMD Line argument strings 
+ * @return int 0 if successful.
+ */
+
+int main(int argc, char *argv[]) {
+     
+    int opt = 0;
+    long num1, num2;                /**< Numeric arguments will be converted to a long int. */
+    num1 = num2 = 0;                /**< Initialize the variables. */
```
Comment (reecetyl) [inline src/main.c:34]: Initialize variables on separate lines

File: test/test_arithmetic.c [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+// Tests for divide function
+void test_divide(void) {
+    CU_ASSERT(divide(6, 3) == 2);         // Positive assertion 
+    CU_ASSERT(divide(600, 200) == 3);     // Larger Positive assertion
+    CU_ASSERT(divide(-10, 2) == -5);      // Negative assertion
+    CU_ASSERT(divide(-5, -5) == 1);       // Negative-identity assertion
+    CU_ASSERT_FALSE(divide(3,2) == 3);    // False Assertion
+    CU_ASSERT(divide(0, 5) == 0);         // Zero Assertion
```
Comment (reecetyl) [inline test/test_arithmetic.c:42]: Add test for division by zero

File: test/test_arithmetic.c [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+// Tests for multiply function
+void test_multiply(void) {
+    CU_ASSERT(multiply(5, 3) == 15);            // Positive assertion 
+    CU_ASSERT(multiply(500, 300) == 150000);    // Larger Positive assertion
+    CU_ASSERT(multiply(-5, 3) == -15);          // Negative assertion
+    CU_ASSERT(multiply(-5, 1) == -5);           // Negative-identity assertion
+    CU_ASSERT_FALSE(multiply(3,2) == 3);        // False Assertion
+    CU_ASSERT(multiply(0, 5) == 0);             // Zero Assertion
```
Comment (reecetyl) [inline test/test_arithmetic.c:32]: Add test for INT32_MAX * 2 (overflow)

**Review by reecetyl — CHANGES_REQUESTED** [review]

File: src/arithmetic.c [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+
+/**
+ * @brief Addition function.
+ * 
+ * @param x First number in addition sequence.
+ * @param y Second number in addition sequence.
+ * @return int Sum of the argument values.
+ */
+
+int add(int x, int y)
```
Comment (reecetyl) [inline src/arithmetic.c:21]: example of what we were talking about:

```c
int add(int32_t x, int32_t y, int32_t *res) {
    int rc = 0;

    if (NULL == res) {
        rc = -1;
        goto end;
    }

    if ((y > 0) && (x > INT32_MAX - y)) {
        rc = -1;
        goto end;
    }

    if ((y < 0) && (x < INT32_MIN - y)) {
        rc = -1;
        goto end;
    }

    *res = x + y;

end:
    return rc;
}
```

**Comment by atldaveo** [comment]
The only thing I wasn't totally rectify was that "*" kept doing a wildcard expansion on my directory files, so I replaced the required input with "x" instead. It's reflected on the print_help function and the contents of main() itself.

I'll make the code to standard barr-c on my next submission while also knocking out the remaining issues.

File: .vscode/settings.json [hunk 21]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (reecetyl) [inline .vscode/settings.json:1]: Add to .gitignore

File: src/main.c [hunk 22]
```diff
@@ -0,0 +1,305 @@
...
+#include <unistd.h>
+#include <stdbool.h>
+#include <errno.h>
+#include <string.h>
+#include "arithmetic.h"
+#include "debug.h"
+#include "helpers.h"
+
+typedef enum
+{
```
Comment (reecetyl) [inline src/main.c:25]: Good!

File: src/main.c [hunk 23]
```diff
@@ -0,0 +1,305 @@
...
+    ADD,
+    SUBTRACT,
+    MULTIPLY,
+    DIVIDE,
+    INVALID
+} Operation;
+    
+Operation checkOperator(const char *operator)
+{
+    if (strcmp(operator, "+") == 0)
```
Comment (reecetyl) [inline src/main.c:35]: Prefer use of strncmp over strcmp (length bounded) where possible

File: src/main.c [hunk 24]
```diff
@@ -0,0 +1,305 @@
...
+    { 
+        return ADD;
+    }
+    else if (strcmp(operator, "-") == 0) 
+    {
+        return SUBTRACT;
+    }
+    else if (strcmp(operator, "x") == 0)
```
Comment (reecetyl) [inline src/main.c:43]: Asterisk (*) rather than x

File: src/main.c [hunk 25]
```diff
@@ -0,0 +1,305 @@
...
+ * @param argv Array of CMD Line argument strings 
+ * @return int 0 if successful.
+ */
+
+int main(int argc, char *argv[]) {
+
+    Operation operation;
+
+    //int opt = 0;
+    long num1, num2;                /**< Numeric arguments will be converted to a long int. */
```
Comment (reecetyl) [inline src/main.c:68]: Initialize everything

Numeric values to 0 or -1 as appropriate

Structs to {0}

Pointers to NULL

File: src/main.c [hunk 26]
```diff
@@ -0,0 +1,305 @@
...
+    char *numEnd1, *numEnd2;        /**< Pointers to the ends of num1 and num2. */ 
+    bool argCheck;                  /**< Check number of CL arguments */
+    bool has_num1, has_num2, opCheck;        /**< Validate num1, num2, and operator */
+
+    errno = 0;                      /**< Reset error number from (errno.h) */
+    num1 = 0;                       /**< Initialize the variables. */
+    num2 = 0;                       /**< Initialize the variables. */    
+
+    // TODO: Right number of arguments
+    if (4 != argc)
```
Comment (reecetyl) [inline src/main.c:81]:
```suggestion
// at top of main, able to be changed easily later..
#define ARGCOUNT 4 
// ...later...
    if (ARGCOUNT != argc) 
```

**Review by reecetyl — CHANGES_REQUESTED** [review]

File: src/main.c [hunk 27]
```diff
@@ -0,0 +1,308 @@
...
+    ADD,
+    SUBTRACT,
+    MULTIPLY,
+    DIVIDE,
+    INVALID
+} Operation;
+    
+Operation checkOperator(const char *operator)
+{
+    if (strncmp(operator, "+") == 0)
```
Comment (reecetyl) [inline src/main.c:35]: Check man page for strncmp, it should take the strings to compare and a length (3 arguments)

**Review by reecetyl — CHANGES_REQUESTED** [review]

File: src/arithmetic.c [hunk 28]
```diff
@@ -0,0 +1,48 @@
...
+int multiply(int32_t x, int32_t y)
+{
+    return x * y;
+}
+
+int divide(int32_t x, int32_t y)
+{
+    if (0 == y)                 /** Prevent division by zero */
+        {
+            return -1;
```
Comment (reecetyl) [inline src/arithmetic.c:25]: Use integer return code and pointer to integer to store result. Currently no way to tell the difference between:
-5 / 5 == -1 (correct)
-5 / 0 == -1 (incorrect)

File: src/main.c [hunk 29]
```diff
@@ -0,0 +1,216 @@
...
+ * operator as to preserve computing resources / more precisely compare the operator in
+ * question.
+ * 
+ * @param operator user input operator on CMD Line 
+ * @return Operation determines which arithmetic operation to execute
+ */
+
+Operation checkOperator(const char *operator)
+{
+    if  (0 != operator && 1 == strlen(operator))
```
Comment (reecetyl) [inline src/main.c:52]: When checking a pointer, compare against NULL rather than 0 for readability

File: src/main.c [hunk 30]
```diff
@@ -0,0 +1,216 @@
...
+    char *numEnd1, *numEnd2;                    /**< Pointers to the ends of num1 and num2. */                               
+    
+    Operation operation;                        
+    errno = 0;                                          
+    /**
+    * @brief Error-checking for number of CMD Line arguments
+    * Should be 4 <filename> <operand> <Operator> <operand>
+    */
+
+    if (ARGCOUNT != argc)
```
Comment (reecetyl) [inline src/main.c:117]: This condition is checked above

File: src/main.c [hunk 31]
```diff
@@ -0,0 +1,216 @@
...
+
+    num1 = strtol(argv[1], &numEnd1, 10);
+    if (0 != errno || '\0' != *numEnd1)     /**< Check for over/underflow and validity of string */
+    {
+        fprintf(stderr, "'%s' is an invalid operand\n", argv[1]);
+        printf("Make sure you input the following:\n");
+        print_help(argv[0]);
+        exit(EXIT_FAILURE);
+    }
+    else
```
Comment (reecetyl) [inline src/main.c:140]: Brackets

File: src/main.c [hunk 32]
```diff
@@ -0,0 +1,216 @@
...
+            case SUBTRACT:
+                result = subtract(num1, num2);
+                printf("%d - %d = %d\n", num1, num2, result);
+                break;
+            case MULTIPLY:
+                result = multiply(num1, num2);
+                printf("%d * %d = %d\n", num1, num2, result);
+                break;
+            case DIVIDE:
+                if (num2 == 0)
```
Comment (reecetyl) [inline src/main.c:202]: Do error handling inside operator function

**Review by reecetyl — CHANGES_REQUESTED** [review]

File: .vscode/settings.json [hunk 33]
```diff
@@ -0,0 +1,13 @@
...
+    "files.associations": {
+        "stdbool.h": "c",
```
Comment (reecetyl) [inline .vscode/settings.json:3]: To remove this file make a local copy, remove from folder, repush

File: include/helpers.h [hunk 34]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * @brief Prints instructions on what the CMD Line input
+ * for simple-calc ought to be. 
+ * 
+ * @param Passes the name of the program (ie. the first CMD Line
+ * input) as argv[0] as a const char pointer so that it can't be
+ * altered. 
+ */
+
+void printHelp(const char *programName);
```
Comment (reecetyl) [inline include/helpers.h:26]: print_help

Comment (reecetyl) [reply include/helpers.h:26]: program_name

File: include/helpers.h [hunk 35]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * @brief Prints the version information for simple-calc.
+ * 
+ */
+
+void printVersion();
```
Comment (reecetyl) [inline include/helpers.h:33]: print_version

File: src/arithmetic.c [hunk 36]
```diff
@@ -0,0 +1,93 @@
...
+ */
+
+#include <stdint.h>
+#include "arithmetic.h"
+#include "debug.h"
+
+int32_t add(int32_t x, int32_t y)
+{
+    // Check for overflow
+    if (0 < x && 0 < y && (INT32_MAX - y) < x)
```
Comment (reecetyl) [inline src/arithmetic.c:19]: Make all comparisons explicitly bracketed 

```suggestion
    if ((0 < x) && (0 < y) && ((INT32_MAX - y) < x))
```

File: src/helpers.c [hunk 37]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+void printHelp(const char *programName)
+{
+    printf("Usage: %s [OPTIONS]\n", programName);
+    printf("Options:\n");
+    printf("  <num1> + <num2>     Add two numbers\n");
+    printf("  <num1> - <num2>     Subtract two numbers\n");
+    printf("  <num1> * <num2>     Multiply two numbers\n");
+    printf("  <num1> / <num2>     Divide two numbers\n");
```
Comment (reecetyl) [inline src/helpers.c:31]: Good, now add left and right shift (<< and >>)

File: src/main.c [hunk 38]
```diff
@@ -0,0 +1,210 @@
...
+    bool hasNum2 = false;
+    bool opCheck = false;        
+    bool argCheck = false;
+    int32_t num1 = 0;                           /**< Variable that will hold an operand with MAX 32 bits */
+    int32_t num2 = 0;                           /**< Variable that will hold an operand with MAX 32 bits */
+    int32_t result = 0;                         /**< Variable that will hold the resulting calculated number */
+    char *numEnd1 = NULL;                       /**< Pointer to the end of num1 */
+    char *numEnd2 = NULL;                       /**< Pointer to the end of num2 */                               
+    
+    Operation operation;
```
Comment (reecetyl) [inline src/main.c:98]:
```suggestion
    Operation operation = {0};                        
```

File: src/main.c [hunk 39]
```diff
@@ -0,0 +1,210 @@
...
+/**
+ * @brief Based on passing error-checks, execute the proper arithmetic
+ * operation and print the results in the form of <num1> <operator> <num2> = <result> 
+ * to the terminal.
+ * 
+ * All the cases for the switch (operation) refer to the enumeration defined at the beginning
+ * of the code. 
+ */
+
+    if (argCheck && hasNum1 && hasNum2 && opCheck)
```
Comment (reecetyl) [inline src/main.c:179]: Make all explicit 

```suggestion
    if ((true == argCheck) && (true == hasNum1) && (true == hasNum2) && (true == opCheck)) 
```

Can also simplify by anding all Boolean's beforehand (only enter condition if all are true)

File: src/main.c [hunk 40]
```diff
@@ -0,0 +1,210 @@
...
+                break;
+            case MULTIPLY:
+                result = multiply(num1, num2);
+                printf("%d * %d = %d\n", num1, num2, result);
+                break;
+            case DIVIDE:
+                if (num2 == 0) 
+                {
+                    fprintf(stderr, "Error: Division by zero is not allowed.\n");
+                    exit(EXIT_FAILURE);
```
Comment (reecetyl) [inline src/main.c:199]: Have a single exit point (set return code and use gotos to go to the bottom)

**Review by reecetyl — CHANGES_REQUESTED** [review]

**Comment by atldaveo** [comment]
Still need to integrate Left / Right shift functionality and establish single exit point for the program

File: src/arithmetic.c [hunk 41]
```diff
@@ -0,0 +1,129 @@
...
+#include "arithmetic.h"
+#include "debug.h"
+
+int32_t add(int32_t x, int32_t y, int32_t *res)
+{
+    // Check for overflow
+    if ((0 < x) && (0 < y) && ((INT32_MAX - y) < x)) 
+    {
+        fprintf(stderr, "\nOverflow: Result is too large\n");
+        return -1;
```
Comment (reecetyl) [inline src/arithmetic.c:22]: Same thing with a single return point at the end of the function.

```suggestion
int rc = -1; // init as failure

// do stuff, if success condition 
rc = 0;
goto end;

// if failure condition 
goto end;

// at bottom
end:
  return rc;
```

**Review by reecetyl — CHANGES_REQUESTED** [review]

File: include/helpers.h [hunk 42]
```diff
@@ -0,0 +1,35 @@
+/**
+ * @file helpers.h
+ * @author David Chung
+ * @brief A series of function prototypes for miscellaneous helper functions
+ * to be used throughout the main program simple-calc.
+ * 
+ * @version 2.1
+ * @date 2024-11-25
+ * 
+ * @copyright Copyright (c) 2024
```
Comment (reecetyl) [inline include/helpers.h:10]: Copyright??

File: src/arithmetic.c [hunk 43]
```diff
@@ -0,0 +1,178 @@
...
+    }
+    else
+    {
+        *res = x / y;       /**< Allows for the result to be stored in memory location &res */
+        return 0;          
+    }
+}
+
+int32_t shift_right(int32_t num, int32_t space, int32_t *res)
```
Comment (reecetyl) [inline src/arithmetic.c:136]: uint32

File: src/arithmetic.c [hunk 44]
```diff
@@ -0,0 +1,178 @@
...
+    }
+    // Upon input validation
+    else
+    {
+        *res = num >> space;
+        return 0;
+    }
+}
+
+int32_t shift_left(int32_t num, int32_t space, int32_t *res)
```
Comment (reecetyl) [inline src/arithmetic.c:158]: uint32

**Review by reecetyl — CHANGES_REQUESTED** [review]
